### PR TITLE
Fix(pinpoint) 2762: Resolve Android 12 notification trampolining restrictions and ADM Push Notifications

### PR DESCRIPTION
**Pull request by tylerjroach**
*Issue #, if available: [2762](https://github.com/aws-amplify/aws-sdk-android/issues/2762)* and [2927](https://github.com/aws-amplify/aws-sdk-android/issues/2927)

*Description of changes:*
Android 12 introduced the restriction of not allowing the "trampolining" of push notifications from a BroadcastReceiver to an Activity. The proposed solution is moving the trampoline effect from a BroadcastReceiver to an Activity. Google's original intent on the restriction was performance concerns by trampolining, where the user could experience a delay between notification click and app launch. We should ensure PinpointNotificationActivity remains light to keep these concerns to a minimum. An alternative involving backstack manipulation was explored, but this option appears less fragile / more simple.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-android-sdk-pinpoint/src/main/java/com/amazonaws/mobileconnectors/pinpoint/targeting/notification/PinpointNotificationActivity.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright 2016-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (sktimalsina) [inline aws-android-sdk-pinpoint/src/main/java/com/amazonaws/mobileconnectors/pinpoint/targeting/notification/PinpointNotificationActivity.java:2]: 2022

File: aws-android-sdk-pinpoint/src/main/java/com/amazonaws/mobileconnectors/pinpoint/targeting/notification/PinpointNotificationActivity.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+public class PinpointNotificationActivity extends Activity {
+    private static String TAG = PinpointNotificationActivity.class.getSimpleName();
+
+    private static volatile NotificationClient notificationClient = null;
+
+    public static void setNotificationClient(NotificationClient notificationClient) {
+        PinpointNotificationActivity.notificationClient = notificationClient;
+    }
+
+    public static void setNotificationClient(NotificationClientBase notificationClientBase) {
```
Comment (sdhuka) [inline aws-android-sdk-pinpoint/src/main/java/com/amazonaws/mobileconnectors/pinpoint/targeting/notification/PinpointNotificationActivity.java:37]: Can we avoid adding static methods in the activity? if the activity is recreated then this property will always be null.

**Review by sdhuka — APPROVED** [review]

**Review by sdhuka — APPROVED** [review]

**Review by tjleing — APPROVED** [review]

**Comment by tylerjroach** [comment]
This PR will also fix issue: [2927](https://github.com/aws-amplify/aws-sdk-android/issues/2927)

File: aws-android-sdk-pinpoint/src/test/java/com/amazonaws/mobileconnectors/pinpoint/targeting/notification/PinpointNotificationActivityTests.java [hunk 3]
```diff
@@ -2,43 +2,47 @@
...
-    private PinpointNotificationReceiver receiver;
-
-    private Intent intent;
-
-    private Context context;
+    private ActivityController<PinpointNotificationActivity> activityController;
 
     @Before
     public void setup() {
+        PinpointNotificationActivity.setNotificationClient((NotificationClient) null);
```
Comment (tjleing) [inline aws-android-sdk-pinpoint/src/test/java/com/amazonaws/mobileconnectors/pinpoint/targeting/notification/PinpointNotificationActivityTests.java:28]: why are we setting this to `null` here before immediately overriding it with a mock?

Comment (tylerjroach) [reply aws-android-sdk-pinpoint/src/test/java/com/amazonaws/mobileconnectors/pinpoint/targeting/notification/PinpointNotificationActivityTests.java:28]: We could remove here. It was just an effort to make sure we first started all tests in a clean slate (ex: this static variable being wiped out), before moving on to additional setup configuration.

**Review by tjleing — APPROVED** [review]

**Review by sktimalsina — APPROVED** [review]

**Review by sdhuka — APPROVED** [review]

**Pull request merged**